src/main/java/com/google/sps/servlets/GetUserData.java

```diff
@@ -0,0 +1,116 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.json.JSONObject;
+
+@WebServlet("/get-user")
+public class GetUserData extends HttpServlet {
+  private FirebaseAuth authInstance;
+  private DatastoreService datastore;
+
+  @Override
+  public void init(ServletConfig config) throws ServletException {
+    try {
+      authInstance = FirebaseAuth.getInstance(FirebaseAppManager.getApp());
+    } catch (IOException e) {
+      throw new ServletException(e);
+    }
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    System.setProperty(
+        DatastoreServiceConfig.DATASTORE_EMPTY_LIST_SUPPORT, Boolean.TRUE.toString());
+
+    try {
+      // Find user ID
+      String idToken = request.getParameter("idToken");
+      FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
+      String userID = decodedToken.getUid();
+      UserRecord userRecord = authInstance.getUser(userID);
+      String userEmail = userRecord.getEmail();
+
+      PreparedQuery queryUser =
+          datastore.prepare(
+              new Query("User")
+                  .setFilter(new FilterPredicate("userEmail", FilterOperator.EQUAL, userEmail)));
+
+      Entity userEntity;
+      // If the student user entity doesnt exist yet, create one
+      if (queryUser.countEntities() == 0) {
+        userEntity = new Entity("User");
+        userEntity.setProperty("userEmail", userEmail);
+        userEntity.setProperty("registeredClasses", Collections.emptyList());
+        userEntity.setProperty("ownedClasses", Collections.emptyList());
+        userEntity.setProperty("taClasses", Collections.emptyList());
+
+        datastore.put(userEntity);
+      } else {
+        userEntity = queryUser.asSingleEntity();
+      }
+
+      List<Key> classList = new ArrayList<Key>();
+      if (request.getParameter("registeredClasses") != null) {
+        classList = (List<Key>) userEntity.getProperty("registeredClasses");
+
+      } else if (request.getParameter("ownedClasses") != null) {
+        classList = (List<Key>) userEntity.getProperty("ownedClasses");
+
+      } else if (request.getParameter("taClasses") != null) {
+        classList = (List<Key>) userEntity.getProperty("taClasses");
+      }
+
+      Map<String, String> classMap = new HashMap<>();
+      for (Key classKey : classList) {
+        String classCode = KeyFactory.keyToString(classKey);
+        String className = (String) datastore.get(classKey).getProperty("name");
+
+        classMap.put(classCode, className);
+      }
+
+      if (classMap.isEmpty()) {
+        response.setContentType("application/json;");
+        response.getWriter().print("null");
+
+      } else {
+        response.setContentType("application/json;");
+        response.getWriter().print(new JSONObject(classMap).toString());
```

Comment:
Is there a reason your using `JSONObject` instead of `Gson`. In the rest of the servlets we use `Gson`, it would be good to be consistent.

Reply:
Couldn't get Gson to return the hashmap the way I wanted (looks like gson handles empty hashmaps differently), so I switched to JSONObject